File: android/src/main/java/app/capgo/rudderstack/RudderStackPlugin.java
package app.capgo.rudderstack;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

@CapacitorPlugin(name = "RudderStack")
public class RudderStackPlugin extends Plugin {

    private static final String VERSION = BuildConfig.VERSION_NAME;

    private RudderStackBridge implementation;

    @Override
    public void load() {
        implementation = new RudderStackBridge(getContext());
    }

    @PluginMethod
    public void initialize(final PluginCall call) {
        final String writeKey = call.getString("writeKey");
        if (writeKey == null || writeKey.trim().isEmpty()) {
            call.reject("writeKey is required");
            return;
        }

        try {
            implementation.initialize(writeKey, call.getObject("config"), call.getObject("options"));
            call.resolve();
        } catch (Exception exception) {
            call.reject(exception.getMessage(), exception);
        }
    }

    @PluginMethod
    public void identify(final PluginCall call) {
        final String userId = call.getString("userId");
        if (userId == null || userId.trim().isEmpty()) {
            call.reject("userId is required");
            return;
        }

        try {
            implementation.identify(userId, call.getObject("traits"), call.getObject("options"));
            call.resolve();
        } catch (Exception exception) {
            call.reject(exception.getMessage(), exception);
        }
    }

    @PluginMethod
    public void group(final PluginCall call) {
        final String groupId = call.getString("groupId");
        if (groupId == null || groupId.trim().isEmpty()) {
            call.reject("groupId is required");
            return;
        }

        try {
            implementation.group(groupId, call.getObject("groupTraits"), call.getObject("options"));
            call.resolve();
        } catch (Exception exception) {
            call.reject(exception.getMessage(), exception);
        }
    }

    @PluginMethod
    public void track(final PluginCall call) {
        final String eventName = call.getString("eventName");
        if (eventName == null || eventName.trim().isEmpty()) {
            call.reject("eventName is required");
            return;
        }

        try {
            implementation.track(eventName, call.getObject("properties"), call.getObject("options"));
            call.resolve();
        } catch (Exception exception) {
            call.reject(exception.getMessage(), exception);
        }
    }

    @PluginMethod
    public void screen(final PluginCall call) {
        final String screenName = call.getString("screenName");
        if (screenName == null || screenName.trim().isEmpty()) {
            call.reject("screenName is required");
            return;
        }

        try {
            implementation.screen(screenName, call.getObject("properties"), call.getObject("options"));
            call.resolve();
        } catch (Exception exception) {
            call.reject(exception.getMessage(), exception);
        }
    }

    @PluginMethod
    public void alias(final PluginCall call) {
        final String newId = call.getString("newId");
        if (newId == null || newId.trim().isEmpty()) {
            call.reject("newId is required");
            return;
        }

        try {
            implementation.alias(newId, call.getObject("options"));
            call.resolve();
        } catch (Exception exception) {
            call.reject(exception.getMessage(), exception);
        }
    }

    @PluginMethod
    public void reset(final PluginCall call) {
        try {
            implementation.reset();
            call.resolve();
        } catch (Exception exception) {
            call.reject(exception.getMessage(), exception);
        }
    }

    @PluginMethod
    public void flush(final PluginCall call) {
        try {
            implementation.flush();
            call.resolve();
        } catch (Exception exception) {
            call.reject(exception.getMessage(), exception);
        }
    }

    @PluginMethod
    public void putDeviceToken(final PluginCall call) {
        final String deviceToken = call.getString("deviceToken");
        if (deviceToken == null || deviceToken.trim().isEmpty()) {
            call.reject("deviceToken is required");
            return;
        }

        implementation.putDeviceToken(deviceToken);
        call.resolve();
    }

    @PluginMethod
    public void setAdvertisingId(final PluginCall call) {
        putAdvertisingId(call);
    }

    @PluginMethod
    public void putAdvertisingId(final PluginCall call) {
        final String advertisingId = call.getString("advertisingId");
        if (advertisingId == null || advertisingId.trim().isEmpty()) {
            call.reject("advertisingId is required");
            return;
        }

        implementation.putAdvertisingId(advertisingId);
        call.resolve();
    }

    @PluginMethod
    public void setAnonymousId(final PluginCall call) {
        putAnonymousId(call);
    }

    @PluginMethod
    public void putAnonymousId(final PluginCall call) {
        final String anonymousId = call.getString("anonymousId");
        if (anonymousId == null || anonymousId.trim().isEmpty()) {
            call.reject("anonymousId is required");
            return;
        }

        implementation.putAnonymousId(anonymousId);
        call.resolve();
    }

    @PluginMethod
    public void optOut(final PluginCall call) {
        final Boolean optOut = call.getBoolean("optOut");
        if (optOut == null) {
            call.reject("optOut is required");
            return;
        }

        try {
            implementation.optOut(optOut);
            call.resolve();
        } catch (Exception exception) {
            call.reject(exception.getMessage(), exception);
        }
    }

    @PluginMethod
    public void getPluginVersion(final PluginCall call) {
        final JSObject ret = new JSObject();
        ret.put("version", VERSION);
        call.resolve(ret);
    }
}

File: ios/Sources/RudderStackPlugin/RudderStackPlugin.swift
import Foundation
import Capacitor

@objc(RudderStackPlugin)
public class RudderStackPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "RudderStackPlugin"
    public let jsName = "RudderStack"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "initialize", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "identify", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "group", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "track", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "screen", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "alias", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "reset", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "flush", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "putDeviceToken", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setAdvertisingId", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "putAdvertisingId", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setAnonymousId", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "putAnonymousId", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "optOut", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getPluginVersion", returnType: CAPPluginReturnPromise)
    ]

    private let implementation = RudderStackBridge()

    @objc func initialize(_ call: CAPPluginCall) {
        guard let writeKey = call.getString("writeKey"), !writeKey.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            call.reject("writeKey is required")
            return
        }

        do {
            try implementation.initialize(
                writeKey: writeKey,
                config: call.getObject("config"),
                options: call.getObject("options")
            )
            call.resolve()
        } catch {
            call.reject(error.localizedDescription)
        }
    }

    @objc func identify(_ call: CAPPluginCall) {
        guard let userId = call.getString("userId"), !userId.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            call.reject("userId is required")
            return
        }

        do {
            try implementation.identify(
                userId: userId,
                traits: call.getObject("traits"),
                options: call.getObject("options")
            )
            call.resolve()
        } catch {
            call.reject(error.localizedDescription)
        }
    }

    @objc func group(_ call: CAPPluginCall) {
        guard let groupId = call.getString("groupId"), !groupId.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            call.reject("groupId is required")
            return
        }

        do {
            try implementation.group(
                groupId: groupId,
                groupTraits: call.getObject("groupTraits"),
                options: call.getObject("options")
            )
            call.resolve()
        } catch {
            call.reject(error.localizedDescription)
        }
    }

    @objc func track(_ call: CAPPluginCall) {
        guard let eventName = call.getString("eventName"), !eventName.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            call.reject("eventName is required")
            return
        }

        do {
            try implementation.track(
                eventName: eventName,
                properties: call.getObject("properties"),
                options: call.getObject("options")
            )
            call.resolve()
        } catch {
            call.reject(error.localizedDescription)
        }
    }

    @objc func screen(_ call: CAPPluginCall) {
        guard let screenName = call.getString("screenName"), !screenName.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            call.reject("screenName is required")
            return
        }

        do {
            try implementation.screen(
                screenName: screenName,
                properties: call.getObject("properties"),
                options: call.getObject("options")
            )
            call.resolve()
        } catch {
            call.reject(error.localizedDescription)
        }
    }

    @objc func alias(_ call: CAPPluginCall) {
        guard let newId = call.getString("newId"), !newId.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            call.reject("newId is required")
            return
        }

        do {
            try implementation.alias(
                newId: newId,
                options: call.getObject("options")
            )
            call.resolve()
        } catch {
            call.reject(error.localizedDescription)
        }
    }

    @objc func reset(_ call: CAPPluginCall) {
        do {
            try implementation.reset()
            call.resolve()
        } catch {
            call.reject(error.localizedDescription)
        }
    }

    @objc func flush(_ call: CAPPluginCall) {
        do {
            try implementation.flush()
            call.resolve()
        } catch {
            call.reject(error.localizedDescription)
        }
    }

    @objc func putDeviceToken(_ call: CAPPluginCall) {
        guard let deviceToken = call.getString("deviceToken"), !deviceToken.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            call.reject("deviceToken is required")
            return
        }

        implementation.putDeviceToken(deviceToken)
        call.resolve()
    }

    @objc func setAdvertisingId(_ call: CAPPluginCall) {
        putAdvertisingId(call)
    }

    @objc func putAdvertisingId(_ call: CAPPluginCall) {
        guard let advertisingId = call.getString("advertisingId"), !advertisingId.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            call.reject("advertisingId is required")
            return
        }

        implementation.putAdvertisingId(advertisingId)
        call.resolve()
    }

    @objc func setAnonymousId(_ call: CAPPluginCall) {
        putAnonymousId(call)
    }

    @objc func putAnonymousId(_ call: CAPPluginCall) {
        guard let anonymousId = call.getString("anonymousId"), !anonymousId.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            call.reject("anonymousId is required")
            return
        }

        implementation.putAnonymousId(anonymousId)
        call.resolve()
    }

    @objc func optOut(_ call: CAPPluginCall) {
        guard let optOut = call.getBool("optOut") else {
            call.reject("optOut is required")
            return
        }

        do {
            try implementation.optOut(optOut)
            call.resolve()
        } catch {
            call.reject(error.localizedDescription)
        }
    }

    @objc func getPluginVersion(_ call: CAPPluginCall) {
        call.resolve([
            "version": implementation.version
        ])
    }
}
